src/rust/catloop/futures/accept.rs

```diff
@@ -0,0 +1,214 @@
+//======================================================================================================================
+
+impl Future for AcceptFuture {
+    type Output = Result<(QDesc, SocketAddrV4, Rc<DuplexPipe>), Fail>;
+
+    /// Polls the target [AcceptFuture].
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
+        let self_: &mut AcceptFuture = self.get_mut();
+
+        // Poll Catmem to make progress on ongoing operations.
+        self_.catmem.borrow_mut().poll();
+
+        // Act according to the state in the connection establishment protocol.
+        match &self_.state {
+            ServerState::WaitingConnectRequest(qt_rx) => listen_and_accept_connections(self_, ctx, *qt_rx),
+            ServerState::AcceptedConnectRequest(qt_tx, duplex_pipe) => {
+                establish_connection(self_, ctx, *qt_tx, duplex_pipe.clone())
+            },
+            ServerState::ConnectionEstablished(new_qd, remote, duplex_pipe) => {
+                Poll::Ready(Ok((*new_qd, *remote, duplex_pipe.clone())))
+            },
+        }
+    }
+}
+
+//======================================================================================================================
+// Standalone Functions
+//======================================================================================================================
+
+// Checks if a connection request is valid.
+fn check_connect_request(catmem: &Rc<RefCell<CatmemLibOS>>, handle: SchedulerHandle, qt: QToken) -> Result<(), Fail> {
+    // Parse and check request.
+    let passed: bool = {
+        let qr: demi_qresult_t = catmem.borrow_mut().pack_result(handle, qt)?;
+        let sga: demi_sgarray_t = match qr.qr_opcode {
+            demi_opcode_t::DEMI_OPC_POP => unsafe { qr.qr_value.sga },
+            _ => panic!("unxpected operation on control duplex pipe"),
+        };
+        let passed: bool = CatloopLibOS::is_magic_connect(&sga);
+        catmem.borrow_mut().free_sgarray(sga)?;
+        passed
+    };
+
+    if !passed {
+        let e: Fail = Fail::new(libc::EPROTO, "invalid request");
+        error!("failed to establish connection ({:?})", e);
+        Err(e)
+    } else {
+        Ok(())
+    }
+}
+
+// Sends the port number to the peer process.
+fn send_port_number(
+    catmem: &Rc<RefCell<CatmemLibOS>>,
+    control_duplex_pipe: Rc<DuplexPipe>,
+    port: u16,
+) -> Result<QToken, Fail> {
+    let sga: demi_sgarray_t = catmem.borrow_mut().alloc_sgarray(mem::size_of_val(&port))?;
+    let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
+    let len: usize = sga.sga_segs[0].sgaseg_len as usize;
+    let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+    slice.copy_from_slice(&port.to_ne_bytes());
+    let qt_tx: QToken = control_duplex_pipe.push(&sga)?;
+    catmem.borrow_mut().free_sgarray(sga)?;
+    Ok(qt_tx)
+}
+
+/// Waits for a connection request to arrive.
+fn listen_and_accept_connections(
```

Comment:
Would naming this just `listen_and_accept` be a good idea?

Reply:
Sure, I renamed it.